api/turing/api/appcontext.go

```diff
@@ -187,23 +186,33 @@ func NewAppContext(
 	return appContext, nil
 }
 
-// getEnvironmentClusterMap creates a map of the environment name to the kubernetes cluster. Additionally,
-// clusters that are not a part of the deployment environments can be registered using the clusterNames
-// parameter (in such cases, the environment name will be saved to be the same as the cluster name).
-func getEnvironmentClusterMap(mlpSvc service.MLPService, clusterNames []string) (map[string]string, error) {
-	envClusterMap := map[string]string{}
+// buildKubeconfigStore creates a map of the environment name to the kubernetes cluster.
+// It combines the EnsemblerServiceBuilderConfig with a list of environments retrieved from mlpSvc
+// into a map. Each environment retrieved from mlpSvc should have a corresponding k8sConfig, else
+// an error is returned.
+func buildKubeconfigStore(mlpSvc service.MLPService, cfg *config.Config) (map[string]*mlpcluster.K8sConfig, error) {
+	// Create a map of env name to cluster name for each supported deployment environment
+	k8sConfigStore := make(map[string]*mlpcluster.K8sConfig)
+	if !cfg.ClusterConfig.InClusterConfig {
+		k8sConfigStore[cfg.EnsemblerServiceBuilderConfig.ClusterName] = cfg.ClusterConfig.EnsemblingServiceK8sConfig
+	} else {
+		k8sConfigStore[cfg.EnsemblerServiceBuilderConfig.ClusterName] = nil
+	}
+	for _, envconfig := range cfg.ClusterConfig.EnvironmentConfigs {
+		k8sConfigStore[envconfig.Name] = envconfig.K8sConfig
+	}
+
 	// Get all environments
 	environments, err := mlpSvc.GetEnvironments()
 	if err != nil {
-		return envClusterMap, err
+		return k8sConfigStore, err
 	}
-	// Create a map of the environment name to cluster id
+	// check if k8s store contains kubeconfig for envs received from merlin
 	for _, environment := range environments {
-		envClusterMap[environment.Name] = environment.Cluster
-	}
-	// Add other required clusters that are not a part of the environments
-	for _, clusterName := range clusterNames {
-		envClusterMap[clusterName] = clusterName
+		// check if clusterConfigs have k8sconfig for environment name
+		if _, ok := k8sConfigStore[environment.Name]; !ok {
+			return nil, fmt.Errorf("Missing k8sconfig for cluster %s", environment.Cluster)
+		}
```

Comment:
Ah okay, that's true, yeah we can just leave this for now :thumbsup: To push the question on decoupling just one bit further, with the current implementation of the Turing API mounting its own copy of the cluster configs (either from its own helm chart or via the MLP chart) onto itself, we can completely remove the dependency of Turing on the `\environments` API altogether in the future right? Like there's no need for it to retrieve the list of environments from any other service (neither Merlin nor MLP)?

Reply:
That is a good question. My current idea is - the MLP API can manage the environments and all the common properties. Also, per the RFC on the Vault dependency removal:

> When deployed via the Caraml umbrella chart, the secret resource name should be set via the global field in values.yaml, so both Merlin and Turing deployment templates can refer to it (this only applies to merlin as of now, since Turing is yet part of caraml-dev/helm-chart repo).

So the environments credentials will be mounted via a common secret created by the CaraML chart, onto Turing, Merlin, etc.

But:

> we can completely remove the dependency of Turing on the \environments API altogether

I agree with this. This is a possibility too. We can consider these options closely when we get to it, whether there is any use to keeping the `/environments` at all.

Reply:
Alright, that's true, we'll come back to this again another time and see the best way to standardise this across the various apps :)